### ISSUE TITLE
Clarification on Server-Sent-Events, persistent connections and concurrency

### ISSUE THREAD

**Issue by keithpitt** [puma/puma#1009]
Howdy!

I've got a quick and (probably stupid) question. I'm currently toying with SSE and Puma (and ActionController::Live). It works like a treat!

```ruby
class SSEController < ActionController::Base
  include ActionController::Live

  def subscriptions
    # SSE expects the `text/event-stream` content type
    response.headers['Content-Type'] = 'text/event-stream'

    # Setup our ActionController::Live Server-Sent-Event Stream
    sse = ActionController::Live::SSE.new(response.stream)

    # Since calling #subscribe locks the redis connection, we need a new one
    # each for each action
    redis = Redis::Namespace.new(REDIS_NAMESPACE, redis: Redis.new(url: REDIS_URL))

    # Create a heartbeat thread that will send a "heartbeat" event to the
    # client every 30 seconds. If this raises an error, we'll know that the
    # client has disconnected.
    heartbeat = Thread.new do
      loop do
        sleep 30
        sse.write ''.freeze, event: 'heartbeat'.freeze
      end
    end

    # Figure out the redis channel we should be listening on
    channel = "anything"

    # The actual thread that will subscribe to redis and publish events
    subscription = Thread.new do
      redis.subscribe(channel) do |on|
        on.message do |channel, data|
          sse.write(data, event: 'data'.freeze)
        end
      end
    end

    # Join the 2 threads back to the main thread
    heartbeat.join
    subscription.join
  rescue ActionController::Live::ClientDisconnected
    # Client disconnected
  ensure
    # All the cleanup
    Thread.kill(heartbeat) if heartbeat
    Thread.kill(subscription) if subscription
    redis.quit if redis
    sse.close if sse
  end
end
```

My question is around how Puma will scale and handle requests to this concurrently. I've noticed that if I specify `workers 1; threads 2 2` in my `puma.rb` configuration, the server can only handle 2 concurrent requests at any one time. This means that if another user tries to load a page while another user is viewing it, it'll just block until the SSE connection closes.

What are some strategies around scaling this sort of setup? Do I just make the `threads` config over 1000 to handle over 1000 users? Is that too high? Am I doing this totally wrong? I found this https://github.com/SamSaffron/message_bus project by @SamSaffron which says it can handle 1000's of concurrent long polling requests (which from my understanding, is the same as SSE, just a prettier API). If I use that and set `threads 2 2` will that have the same problem?

I did some googling and found that Passenger has this `passenger_force_max_concurrent_requests_per_process` config that you can turn on for your `SSE` route - giving it unlimited concurrent requests. It seems like that's their solution to the problem, but I'm wondering what I should with Puma?

Thanks in advance!

**Comment by SamSaffron** [comment]
Puma support Rack Hijack. 

MessageBus and Action Cable build on this. Meaning the request hits puma, and then the hijack callback is made and it "frees" up the slot giving you back a raw socket to do what you will with. In message buses case it has a timerthread that manages 100s or thousands of these. 

This SSE approach you are demonstrating is not using hijack so you will have trouble scaling it. Now ... personally I see zero point in using SSE.

http://caniuse.com/#feat=eventsource

You can not use SSE in IE etc but can use XHR2 http://caniuse.com/#feat=xhr2 which gives you the ability to take advantage of chunked encoding in XHR requests (and stream data through an XHR - which is what message bus does) 

Additionally if you want anything to be robust you got to support fallbacks (for example message bus has non chunked encoding and non long polling fallbacks) 

Bottom line is avoid ActionController::Live, always use rack hijack or libraries building on hijack if you want to scale.

**Comment by TiagoCardoso1983** [comment]
https://gist.github.com/TiagoCardoso1983/591d48cfde04219f801fa7fd7966571c 

Done with sinatra, but the principle is the same, and it works with puma.

**Comment by keithpitt** [comment]
@TiagoCardoso1983 oh great, thanks for the sample code! I'm starting to understand now. By using `rack.hijack`, it frees the connection from Puma allowing other code to keep the connection open and do with it what it wills.

So why doesn't Puma do that sort of thing out of the box? It seems like a good idea! (Sorry again if I'm asking silly questions)

**Comment by evanphx** [comment]
@keithpitt Do which part out of the box?

**Comment by keithpitt** [comment]
@evanphx actually, ignore my comment on that. Reading back on everything it doesn't make sense. But thanks for reaching out for clarification!

I think my confusion came from the `threads` setting. I was looking at how ActionCable and MessageBus work, and from what I saw it stored the connection, returned a response, and managed it in another thread. But I was like "but `threads` is set to 2, how can they use more?! how are they circumventing the Puma setting" I realise that was a bit silly. I think my comment was "well, if I can just use rack.highjack and handled more concurrent requests, why have the `threads` option to begin with?". So it sounds like `threads` is actually more like `concurrent_requests` for that worker?

@SamSaffron thanks for going into more detail about how it all works! I was trying to rack (pardon the pun) my head around how you're able to "free" the thread from Puma. I was looking at SSE mostly because it seemed easy, but you're right, I should definitely look into the other standards.

**Comment by keithpitt** [comment]
(I closed the issue since it's not an "issue" anymore)

**Comment by khataev** [comment]
> https://gist.github.com/TiagoCardoso1983/591d48cfde04219f801fa7fd7966571c
> 
> Done with sinatra, but the principle is the same, and it works with puma.

can you please present this gist once again, I'm curious to see it! Thanks

**Comment by HoneyryderChuck** [comment]
https://gist.github.com/HoneyryderChuck/591d48cfde04219f801fa7fd7966571c
